### ISSUE TITLE
Create Framework #1

### ISSUE THREAD

**Issue by Xz51xt** [Xz51xt/LiquidMuse#1]
# LiquidMuse — Plan From Scratch (État de l'Art 2025-2026)

**Règle absolue : Zéro modèle pré-entraîné, zéro poids importés. PyTorch uniquement comme base.**

---

## Contexte & Objectif

Construire un framework de génération musicale complète (instruments + voix, tous genres, jusqu'à 4-5 minutes) à partir d'un prompt texte. Chaque composant est implémenté et entraîné from scratch. L'architecture s'appuie sur les décisions validées par la recherche 2025-2026 : **VAE de compression haute + Flow Matching + Backbone hybride CfC/Liquid-S4 + Vocodeur Vocos**.

---

## Décisions architecturales clés (justifiées par la recherche)

**Pourquoi Flow Matching plutôt que diffusion classique ou autorégressif ?** La recherche 2025 (DiffRhythm 2, Stable Audio 2) confirme que le Flow Matching (Conditional Flow Matching / Rectified Flow) converge plus vite que la diffusion DDPM classique, génère en moins de steps, et produit une meilleure cohérence structurelle longue durée que les modèles autorégressifs purs. Les modèles autorégressifs ont une légère supériorité perceptuelle mais sont 15× à 50× plus lents à l'inférence.

**Pourquoi Liquid-S4 plutôt que CfC pur ?** Liquid-S4 combine LTC (Liquid Time-Constant) et S4 (Structured State Space), donnant un backbone input-adaptatif avec les propriétés "liquides" ET la puissance des SSM. Il atteint 87.32% sur le Long-Range Arena benchmark et surpasse S4 pur sur la reconnaissance de parole audio brut. Pour la musique (longues dépendances multi-échelles), c'est objectivement supérieur à un CfC seul.

**Pourquoi Vocos plutôt que HiFi-GAN ?** Vocos génère directement des coefficients de Fourier via iSTFT, sans convolutions transposées. Résultat : vitesse d'inférence 10× à 70× supérieure, meilleure qualité sur les métriques VISQOL et PESQ, et élimination des artefacts de périodicité des vocodeurs time-domain.

**Pourquoi VAE continu plutôt que RVQ discret ?** La recherche 2025 (Stable Audio 2, DiffRhythm 2, ACE-Step) converge vers des VAE continus à très bas frame-rate (10-22 Hz) pour le flow matching. Les RVQ discrets (75 Hz) sont compatibles avec l'autorégressif mais créent une séquence trop longue pour le flow matching. À 10-22 Hz, une minute de musique = 600-1320 frames seulement, ce qui rend la génération d'un morceau entier computationnellement raisonnable.

---

## Composants à construire (dans cet ordre strict)

---

### Composant 1 — VAE Audio (Deep Compression)

**But :** Compresser un MEL spectrogram stéréo 44.1kHz en représentation latente continue très compressée, à ~10-22 Hz de frame-rate latent.

**Ce qu'il faut implémenter :**

Un auto-encodeur variationnel (VAE) avec encodeur et décodeur entièrement convolutionnels. L'encodeur prend en entrée le MEL spectrogram et applique des blocs convolutifs successifs avec strides très agressifs dans la dimension temporelle (compression ×32 à ×64 en temps, ×8 en fréquence), aboutissant à un espace latent de dimension 8 à 16 canaux. Le décodeur remonte symétriquement vers le MEL spectrogram complet. L'espace latent est normalisé (moyenne 0, variance 1) via la reparameterization trick du VAE standard.

Le décodeur MEL → waveform est géré séparément par le vocodeur (Composant 4). Le VAE ne produit que le MEL.

**Loss d'entraînement :** reconstruction L1 sur le MEL + terme KL de régularisation latente avec coefficient progressif (annealing de 0 à 1e-4) + discriminateur multi-échelle sur le MEL (patch discriminator) pour forcer la netteté spectrale.

**Paramètres cibles :** frame-rate latent de ~10-22 Hz (compression temporelle ×32 à ×64 sur 44.1kHz), 8 canaux latents. Audio stéréo 44.1kHz.

**Données :** FMA Medium + MTG-Jamendo (audio brut uniquement).

---

### Composant 2 — Encodeur Texte (Transformer from scratch)

**But :** Convertir un prompt texte en vecteurs d'embedding pour le conditionnement.

**Ce qu'il faut implémenter :**

Un tokeniseur BPE entraîné sur un corpus de descriptions musicales (métadonnées FMA, Jamendo : genre, tags, BPM, mood, instruments). Puis un Transformer encoder from scratch (6 couches, 8 têtes, dimension 512, FFN dimension 2048, positional encoding sinusoïdal, LayerNorm pre-norm). Entraîné sur Masked Language Modeling (MLM) avec 15% de tokens masqués.

Ajouter un alignement CLAP-style (text-audio contrastif) : entraîner simultanément un encodeur audio simple (CNN sur MEL) et l'encodeur texte à rapprocher leurs embeddings pour des paires texte-audio correspondantes via InfoNCE loss. Cela améliore drastiquement la correspondance texte-musique à l'inférence.

**Paramètres cibles :** vocab 8k tokens, max sequence 128 tokens, embedding dim 512.

---

### Composant 3 — Backbone Liquid-S4 Multi-Échelle

**But :** Modéliser les dépendances temporelles longues du latent audio à plusieurs vitesses.

**Ce qu'il faut implémenter :**

Liquid-S4 est une combinaison de LTC (Liquid Time-Constant Networks) avec S4 (Structured State Space Models). La dynamique du système est : h(t) = A_liquid(x) * h(t-1) + B * x(t), où A_liquid est une matrice d'état apprise qui dépend de l'entrée (propriété "liquide"), contrairement au S4 standard où A est fixe. Cette dépendance input-adaptative est la différence clé avec un SSM classique.

En pratique, implémenter Liquid-S4 comme un kernel convolutionnel dont les paramètres sont modulés par l'entrée. Le noyau est calculé comme un kernel LTC linéarisé (approximation convolutional de l'ODE LTC), puis appliqué efficacement via FFT.

Empiler 3 niveaux avec des configurations différentes :
- Niveau global (64 états d'état) : contexte complet du morceau, capture structure couplet/refrain
- Niveau moyen (128 états) : phrases musicales, progressions harmoniques
- Niveau fin (64 états) : détails rythmiques, timbre

À chaque niveau, injecter les embeddings texte via cross-attention from scratch (Q depuis le latent, K/V depuis le texte). Utiliser des RoPE embeddings (Rotary Position Embeddings) pour le positionnement.

Le backbone reçoit le latent bruité (pendant le flow matching) + l'encodage du time step t + les embeddings texte, et prédit le champ de vitesse (velocity field) pour le flow matching.

---

### Composant 4 — Vocodeur Vocos (from scratch)

**But :** Convertir un MEL spectrogram en waveform audio de haute qualité.

**Ce qu'il faut implémenter :**

Architecture Vocos : le générateur n'utilise PAS de convolutions transposées. Il maintient la même résolution temporelle dans toutes ses couches internes, puis génère les coefficients de Fourier (partie réelle et imaginaire du STFT) directement. La reconstruction audio se fait via une simple iFFT. Le backbone interne est une pile de blocs ConvNeXt 1D (depthwise conv + inverted bottleneck + LayerNorm + GELU), qui traitent les features MEL sans changer la résolution temporelle.

Discriminateurs : Multi-Period Discriminator (MPD) opérant sur le waveform, et Multi-Resolution STFT Discriminator (MRSD) opérant sur différentes résolutions spectrales.

**Loss :** adversarial GAN loss + feature matching loss + reconstruction L1 sur le MEL.

**Entraînement :** d'abord avec loss de reconstruction seule (100 steps), puis ajouter le discriminateur. Cela stabilise l'entraînement GAN.

---

### Composant 5 — Flow Matching (Generative Process)

**But :** Entraîner le processus génératif qui transforme du bruit gaussien en latent audio conditionné.

**Ce qu'il faut implémenter :**

Conditional Flow Matching (CFM) avec Rectified Flows. Le principe : apprendre un champ de vitesse v_θ(z_t, t, c) qui transporte une distribution simple (bruit gaussien z_0) vers la distribution cible (latents VAE z_1) via un ODE : dz/dt = v_θ(z_t, t, c). Pendant l'entraînement, on construit des trajectoires linéaires droites entre z_0 et z_1, et on supervise v_θ à prédire la direction (z_1 - z_0) à chaque point t de la trajectoire. La loss est simplement MSE entre la vitesse prédite et la vitesse cible (z_1 - z_0).

c est le conditionnement : embeddings texte + time step encodé (sinusoïdal). Classifier-Free Guidance : dropper le prompt texte avec probabilité 15% pendant l'entraînement (remplacer par vecteur nul). À l'inférence, combiner guidance conditionné et non-conditionné : v_guided = v_uncond + scale * (v_cond - v_uncond).

**Inférence :** résoudre l'ODE de t=0 à t=1 avec un solver Euler simple (20-50 steps). Pas besoin de solver complexe grâce aux trajectoires droites des Rectified Flows.

Le backbone du flow matching EST le Liquid-S4 multi-échelle (Composant 3).

---

### Composant 6 — Module Voix & Chant (from scratch)

**But :** Générer une piste vocale chantée conditionnée sur paroles + mélodie.

**Ce qu'il faut implémenter :**

Un VAE dédié à la voix (même architecture que Composant 1 mais entraîné uniquement sur des données de chant), avec un espace latent séparé. Un modèle de chant basé sur Liquid-S4 (architecture légère, 4 couches) qui reçoit en entrée les phonèmes (converti depuis les paroles via un dictionnaire graphème-phonème entraîné), la courbe de pitch F0 (extraite de la piste musicale par l'algorithme PYIN implémenté from scratch), et génère les latents vocaux via Flow Matching conditionné.

Pour la disentanglement voix/style : entraîner avec un encodeur de contenu (phonèmes) séparé d'un encodeur de timbre (style vocal). À l'inférence, le timbre est contrôlé par le prompt ("voix féminine soul grave", "ténor lyrique", "rap masculin sec").

**Données :** GTSinger + NUS-48E + OpenSinger (audio brut de chant avec paroles alignées).

**Mixage voix + musique :** le latent musical et le latent vocal sont générés indépendamment puis fusionnés dans l'espace waveform avec le vocodeur Vocos après décodage séparé des deux VAE.

---

### Composant 7 — Training Pipeline

**Données brutes (licence libre, pas de poids) :**
- FMA Medium — 25k morceaux, 16 genres (audio brut)
- MTG-Jamendo — 55k morceaux Creative Commons (audio brut)
- MAESTRO — piano classique aligné MIDI (audio brut)
- GTSinger + NUS-48E + OpenSinger — pour le module voix (audio brut)

**Preprocessing :** resample à 44.1kHz stéréo, normalisation peak à -1dBFS, découpage en segments de 30 secondes avec overlap 5 secondes, calcul MEL spectrogram from scratch via torch.stft (n_fft=2048, hop=512, n_mels=128), tokenisation des métadonnées en prompts texte automatiques.

**Ordre d'entraînement :**
1. Tokeniseur BPE + alignement text-audio contrastif (Composant 2)
2. VAE audio (Composant 1) — doit être stable avant tout le reste
3. Vocodeur Vocos (Composant 4) — peut s'entraîner en parallèle du VAE
4. Backbone Liquid-S4 + Flow Matching sur latents VAE (Composants 3 + 5)
5. VAE voix + modèle de chant (Composant 6) en parallèle de l'étape 4
6. Fine-tuning global end-to-end (optionnel, avec lr très faible)

**Config entraînement Flow Matching :** AdamW lr=1e-4, warmup 1000 steps, cosine decay, gradient clipping 1.0, bf16 mixed precision, gradient accumulation 4 steps, batch effectif 32. EMA (Exponential Moving Average) des poids avec decay 0.999 pour la stabilité.

---

### Composant 8 — Interface

**CLI :** commande unique avec paramètres : prompt, durée (max 300 secondes), with_vocals, lyrics, voice_style, temperature, cfg_scale, num_inference_steps, seed, output_path.

**API REST (FastAPI) :** endpoint POST /generate avec les mêmes paramètres, retourne le fichier audio WAV en streaming. Endpoint GET /health. Génération async avec queue pour les requêtes longues.

---

## Dépendances (outils uniquement, pas de modèles)

```
torch, torchaudio         — deep learning
numpy, scipy              — calcul scientifique
librosa                   — preprocessing audio offline (PYIN pour F0, etc.)
soundfile                 — I/O audio
fastapi, uvicorn          — API REST
omegaconf                 — gestion config YAML
wandb                     — monitoring
accelerate                — multi-GPU / mixed precision
tqdm                      — progress bars
```

---

## Points critiques pour l'agent

**Sur le VAE :** le frame-rate latent cible est 10-22 Hz. Plus la compression est agressive, plus la génération est rapide, mais il y a un seuil en dessous duquel la qualité se dégrade (la recherche ACE-Step 2025 montre que ×32 = 10.77 Hz est viable, ×64 est trop agressif). Viser ×32 en temps, soit ~10-11 Hz de frame-rate latent pour 44.1kHz.

**Sur Liquid-S4 :** l'implémentation key est que la matrice A (matrice d'état SSM) soit input-dépendante. L'approche la plus stable : calculer A_liquid = A_base + delta_A(input), où delta_A est un petit réseau de neurones. A_base est la matrice S4 diagonale classique. Cela est strictement plus expressif que S4 pur tout en restant calculable via FFT en convolution.

**Sur le Flow Matching :** utiliser des trajectoires linéaires droites (Rectified Flows, z_t = t * z_1 + (1-t) * z_0). Ne pas utiliser les trajectoires cosinusoïdales DDPM — elles sont suboptimales pour le flow matching. La loss target à chaque step est la direction (z_1 - z_0), pas le bruit epsilon.

**Sur Vocos :** l'output du générateur est de dimension (batch, 2, T, n_fft//2+1) — partie réelle et imaginaire du STFT. L'iFFT est appliquée sur cette sortie pour obtenir le waveform. C'est la différence fondamentale avec HiFi-GAN qui prédit les samples temporels directement.

**Sur la génération longue durée :** avec un frame-rate latent de ~11 Hz, 4 minutes = 2640 frames latents. C'est gérable pour le Liquid-S4 (séquences longues = son point fort). Pas besoin de découpage en chunks pour des durées inférieures à 5 minutes.

**Sur le conditionnement texte :** l'alignement CLAP-style (contrastif texte-audio) est essentiel pour que les prompts soient bien suivis. Sans cet alignement, l'encodeur texte entraîné uniquement sur MLM ne "comprend" pas le contenu audio. C'est la leçon principale d'ACE-Step (2025) qui utilise REPA pour forcer cet alignement sémantique.

**Sur la voix :** la synthèse vocale de chant est le composant le plus difficile. La clé est de bien disentangler contenu phonétique (ce qui est chanté) et timbre (comment c'est chanté). Si le disentanglement échoue, la voix générée sonnera incohérente. Prévoir du temps pour débugger ce module spécifiquement.
